### PR TITLE
Add support for reading exit codes on shells

### PR DESCRIPTION
**Pull request by angrybayblade**

### CODE REVIEW

**Comment by utkarsh-dixit** [comment]
### Changes Summary for PR #268

**Objective**: The objective of this PR is to enhance the shell tools by adding support for reading exit codes from shell processes and incorporating these exit codes into the responses of various shell commands.

**Changes Introduced**:
1. **Shell Exit Codes**: Introduced methods to capture and return exit codes from shell processes.
2. **Command Responses**: Modified various command response classes to include exit codes.

**Files Modified**:
- `python/composio/tools/env/docker/shell.py`
- `python/composio/tools/env/host/shell.py`
- `python/composio/tools/local/shelltool/file_cmds/actions/base_file_cmds.py`
- `python/composio/tools/local/shelltool/file_cmds/actions/edit_cmd.py`
- `python/composio/tools/local/shelltool/file_cmds/actions/scroll_cmds.py`
- `python/composio/tools/local/shelltool/find_cmds/actions/search_cmds.py`
- `python/composio/tools/local/shelltool/git_cmds/actions/clone_github.py`
- `python/composio/tools/local/shelltool/git_cmds/actions/get_patch.py`
- `python/composio/tools/local/shelltool/git_cmds/actions/git_tree.py`
- `python/composio/tools/local/shelltool/shell_exec/actions/exec.py`
- `python/composio/tools/local/shelltool/utils.py`

**Category**: Feature

#### Important Change Files:
1. `python/composio/tools/env/docker/shell.py`
2. `python/composio/tools/env/host/shell.py`
3. `python/composio/tools/local/shelltool/file_cmds/actions/base_file_cmds.py`
4. `python/composio/tools/local/shelltool/shell_exec/actions/exec.py`

File: python/composio/tools/env/docker/shell.py [hunk 1]
```diff
@@ -160,6 +160,11 @@ def _communicate_with_handling(self, cmd: str, error: str) -> str:
             raise RuntimeError(f"{cmd}: {error}: {output}")
         return output["stdout"]
 
+    def _get_exit_code(self) -> int:
+        """Get exit code of the last process."""
+        self._write("echo $#")
```
Comment (shubhras01) [inline python/composio/tools/env/docker/shell.py:165]: this should be "echo $?" not "$#", $# gives number of output printed, #? gives correct exit_status of last command executed

File: python/composio/tools/env/host/shell.py [hunk 2]
```diff
@@ -28,13 +33,42 @@ def __init__(self) -> None:
             text=True,
             bufsize=1,
         )
-        self.logger.debug(f"Initial data from session: {self.id} - {self._read()}")
+        self.logger.debug(
+            f"Initial data from session: {self.id} - {self._read(wait=False)}"
+        )
+
+    def _wait_for_cmd(self, cmd: str) -> None:
```
Comment (shubhras01) [inline python/composio/tools/env/host/shell.py:40]: where is this being used ?

**Review by shubhras01 — CHANGES_REQUESTED** [review]

File: python/composio/tools/env/docker/shell.py [hunk 3]
```diff
@@ -200,16 +205,23 @@ def setup(self) -> None:
...
             # TODO: Handle this as framework error
             raise RuntimeError(str(e)) from e
-        return self._read()
+
+    def exec(self, cmd: str) -> t.Dict:
+        """Execute command on container."""
+        self._write(cmd=cmd)
+        return {
+            **self._read(),
+            "exit_code": self._get_exit_code(),
```
Comment (shubhras01) [inline python/composio/tools/env/docker/shell.py:223]: exit-code should be a constant defined above, as we are using this key at other places as well

File: python/composio/tools/env/host/shell.py [hunk 4]
```diff
@@ -28,13 +33,42 @@ def __init__(self) -> None:
...
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            ).stdout.decode()
+            if all(_cmd.lstrip().rstrip() not in output for _cmd in cmd.split("&&")):
+                return
+            time.sleep(1)
+
+    def _get_exit_code(self) -> int:
+        """Get exit code of the last process."""
+        self._write("echo $#")
```
Comment (shubhras01) [inline python/composio/tools/env/host/shell.py:54]: this echo ## should also be a ocnstant --> ECHO_EXIT_STATUS command

File: python/composio/tools/env/host/shell.py [hunk 5]
```diff
@@ -28,13 +33,42 @@ def __init__(self) -> None:
...
+        timeout: float = 120.0,
+    ) -> t.Dict:
         """Read data from a subprocess with a timeout."""
         stderr = t.cast(t.IO[str], self._process.stderr).fileno()
         stdout = t.cast(t.IO[str], self._process.stdout).fileno()
         buffer = {stderr: b"", stdout: b""}
+        if wait and cmd is None:
+            raise ValueError("`cmd` cannot be `None` when `wait` is set to `True`")
+
+        if wait:
```
Comment (shubhras01) [inline python/composio/tools/env/host/shell.py:70]: what is the point of having "wait" and "timout" both ?

File: python/composio/tools/env/host/shell.py [hunk 6]
```diff
@@ -74,7 +117,13 @@ def exec(self, cmd: str) -> t.Dict:
             # TODO: Handle this as framework error
             raise RuntimeError(str(e)) from e
 
-        return self._read()
+    def exec(self, cmd: str) -> t.Dict:
+        """Execute command on container."""
+        self._write(cmd=cmd)
+        return {
+            **self._read(cmd=cmd, wait=True),
+            "exit_code": self._get_exit_code(),
```
Comment (shubhras01) [inline python/composio/tools/env/host/shell.py:125]: "exit_code" should be constant

**Review by shubhras01 — CHANGES_REQUESTED** [review]

File: python/composio/tools/env/host/shell.py (hunk 2 above)
Comment (angrybayblade) [reply python/composio/tools/env/host/shell.py:40]: https://github.com/ComposioHQ/composio/blob/879ae3dbc2c9a8eec922a76e6c523436eca01c42/python/composio/tools/env/host/shell.py#L71

File: python/composio/tools/env/host/shell.py (hunk 5 above)
Comment (angrybayblade) [reply python/composio/tools/env/host/shell.py]: Fixed - https://github.com/ComposioHQ/composio/pull/268/commits/89336b1b091e40b78a9d4e862ce9c35b9ab1fdd6#diff-4c160bb7df37eb11849385599ab5110fcf0628e2ed1c7a0950e0365b90fa5bfaR70-R72

**Comment by kaavee315** [comment]
/review
--pr_reviewer.num_code_suggestions="8"

File: python/tests/test_tools/test_local/test_workspace.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
+"""Test workspace tools."""
```
Comment (kaavee315) [inline python/tests/test_tools/test_local/test_workspace.py:1]: Awesome

File: python/composio/tools/local/shelltool/utils.py [hunk 8]
```diff
@@ -21,35 +18,28 @@ def git_reset_cmd(commit_id) -> str:
...
+    """Commands to clone github repository."""
+    *_, reponame = request_data.repo_name.lstrip().rstrip().split("/")
+    github_access_token = os.environ.get("GITHUB_ACCESS_TOKEN")
+    if not github_access_token or not github_access_token.strip():
+        if os.environ.get("ALLOW_CLONE_WITHOUT_REPO") != "true":
+            raise RuntimeError(
+                "Cannot clone github repository without github access token"
+            )
+        commands = [
+            f"git clone --progress https://github.com/{request_data.repo_name}.git",
```
[the remark this answers is not in the document]
Comment (angrybayblade) [reply python/composio/tools/local/shelltool/utils.py:30]: Yes, other wise we won't be able to read the output continuously and might miss out on some of the output logs

**Review by shubhras01 — APPROVED** [review]
lgtm

**Pull request merged**